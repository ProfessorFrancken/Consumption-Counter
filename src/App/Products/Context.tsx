import React from "react";
import {makeOrder as makeOrderAction, TYPES} from "actions";
import {MemberType} from "App/Members/Members";
import {useDispatch, useSelector} from "react-redux";
import {useHistory} from "react-router";
import {ProductPropType} from "./Products";
import {productsWithOrderCountSelector} from "./selectors";

export type Product = {
  id: number;
  name: string;
  price: number;
  position: number;
  category: "Bier" | "Fris" | "Eten";
  image: string;
  splash_image: string;
  age_restriction: number | null;
};

function didNotRecentlyOrderAProduct(member: MemberType) {
  const latest_purchase_at =
    typeof member.latest_purchase_at === "string"
      ? new Date(member.latest_purchase_at)
      : member.latest_purchase_at;

  if (latest_purchase_at === null) {
    return true;
  }

  const today = new Date();
  const timeDiff = today.getTime() - latest_purchase_at.getTime();
  const diffDays = Math.ceil(timeDiff / (1000 * 3600 * 24));

  return diffDays > 90;
}

type Order = {
  products: Product[];
  member: undefined | MemberType;
};

type OrderAction =
  | {type: "TOGGLE_BUY_MORE_PRODUCTS"; product: Product}
  | {type: "SELECT_MEMBER"; member: MemberType}
  | {type: "ADD_PRODUCT_TO_ORDER"; product: Product}
  | {type: "QUEUE_ORDER"};

const emptyOrder: Order = {
  member: undefined,
  products: [],
};

type State = {
  products: Product[];
  productsWithHour: {
    Bier: ProductPropType[];
    Fris: ProductPropType[];
    Eten: ProductPropType[];
  };
  selectMember: (member: MemberType) => void;
  addProductToOrder: (product: Product) => void;
  toggle: (product: Product) => void;
  buyAll: () => void;
  order: Order;
};

const orderReducer = (state: Order, action: OrderAction) => {
  switch (action.type) {
    case "TOGGLE_BUY_MORE_PRODUCTS":
      return {
        ...state,
        products: state.products.length === 0 ? [action.product] : [],
      };
    case "SELECT_MEMBER":
      return {...emptyOrder, member: action.member};
    case "ADD_PRODUCT_TO_ORDER":
      return {...state, products: [...state.products, {...action.product}]};
    case "QUEUE_ORDER":
      return emptyOrder;
  }
};

const ProductPurchaseContext = React.createContext<State | undefined>(undefined);

export const ProductPurchaseProvider: React.FC<{order?: Order}> = ({
  order: defaultOrder = emptyOrder,
  ...props
}) => {
  const dispatch = useDispatch();
  const [order, orderDispatch] = React.useReducer(orderReducer, defaultOrder);

  const {push} = useHistory();

  const products = useSelector((state: any) => state.products);
  const hour = new Date().getHours();

  const productsWithHour = useSelector((state: any) =>
    productsWithOrderCountSelector(state, {hour})
  ) as {
    Bier: ProductPropType[];
    Fris: ProductPropType[];
    Eten: ProductPropType[];
  };

  const makeOrder = (order: Order) => {
    orderDispatch({type: "QUEUE_ORDER"});
    dispatch(makeOrderAction(order));
  };

  const addProductToOrder = (product: any) => {
    if (order.products.length === 0) {
      return makeOrder({member: order.member, products: [product]});
    } else {
      orderDispatch({type: "ADD_PRODUCT_TO_ORDER", product});
      dispatch({type: TYPES.ADD_PRODUCT_TO_ORDER, product});
    }
  };

  const selectMember = (member: MemberType) => {
    if (didNotRecentlyOrderAProduct(member)) {
      if (!window.confirm(`Are you sure you want to select ${member.fullname}?`)) {
        // Cancel the selection since selecting this member was a mistake
        return;
      }
    }

    push("/products");
    orderDispatch({type: "SELECT_MEMBER", member});
    dispatch({type: TYPES.SELECT_MEMBER, member});
  };

  const toggle = (product: Product) => {
    orderDispatch({type: "TOGGLE_BUY_MORE_PRODUCTS", product});
    dispatch({type: TYPES.BUY_MORE, product});
  };

  const value = {
    products,
    productsWithHour,
    selectMember,
    addProductToOrder: (product: Product) => addProductToOrder(product),
    toggle: toggle,
    buyAll: () => makeOrder(order),
    order,
  };

  return <ProductPurchaseContext.Provider value={value} {...props} />;
};

export const useProductPurchase = () => {
  const context = React.useContext(ProductPurchaseContext);

  if (!context) {
    throw new Error(`useProductPurchase must be used within a ProductPurchaseProvider`);
  }

  return context;
};
