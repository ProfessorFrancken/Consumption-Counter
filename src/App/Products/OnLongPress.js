import React, {Component} from "react";
import PropTypes from "prop-types";

class OnLongPress extends Component {
  state = {
    longPressed: false,
    longPressTimeout: null,
  };

  constructor(props) {
    super(props);

    this.start = this.start.bind(this);
    this.end = this.end.bind(this);
    this.onLongPress = this.onLongPress.bind(this);
  }

  start(e) {
    e.preventDefault();

    const longPressTimeOut = setTimeout(this.onLongPress, this.props.timeout);

    this.setState({longPressTimeOut, longPressed: false});
  }

  end(e) {
    e.preventDefault();

    if (this.state.longPressed) {
      this.setState({longPressed: false, longPressTimeOut: null});
      return;
    }

    clearTimeout(this.state.longPressTimeOut);

    this.props.onClick();
  }

  onLongPress() {
    clearTimeout(this.state.longPressTimeOut);

    this.setState({longPressed: true, longPressTimeOut: null});

    this.props.onLongPress();
  }

  render() {
    const {children} = this.props;

    const childrenWithProps = React.Children.map(children, (child) =>
      React.cloneElement(child, {
        onMouseDown: this.start,
        onTouchStart: this.start,
        onMouseUp: this.end,
        onTouchCancel: this.end,
        onTouchEnd: this.end,
      })
    );

    return childrenWithProps;
  }
}

OnLongPress.propTypes = {
  timeout: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  onLongPress: PropTypes.func.isRequired,
};

OnLongPress.defaultProps = {
  timeout: 500,
};

export default OnLongPress;
