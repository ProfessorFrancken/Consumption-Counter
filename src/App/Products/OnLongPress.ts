import React, {Component} from "react";
import PropTypes from "prop-types";
class OnLongPress extends Component {
  static propTypes = {
    timeout: PropTypes.number.isRequired,
    children: PropTypes.node.isRequired,
    onClick: PropTypes.func.isRequired,
    onLongPress: PropTypes.func.isRequired,
  };
  static defaultProps = {
    timeout: 500,
  };
  state = {
    longPressed: false,
    longPressTimeout: null,
  };
  constructor(props: any) {
    super(props);
    this.start = this.start.bind(this);
    this.end = this.end.bind(this);
    this.onLongPress = this.onLongPress.bind(this);
    this.onClick = this.onClick.bind(this);
  }
  start(e: any) {
    e.preventDefault();
    const longPressTimeOut = setTimeout(this.onLongPress, (this.props as any).timeout);
    this.setState({longPressTimeOut, longPressed: false});
  }
  end(e: any) {
    e.preventDefault();
    if (this.state.longPressed) {
      this.setState({longPressed: false, longPressTimeOut: null});
      return;
    }
    // @ts-expect-error ts-migrate(2551) FIXME: Property 'longPressTimeOut' does not exist on type... Remove this comment to see the full error message
    clearTimeout(this.state.longPressTimeOut);
    (this.props as any).onClick();
  }
  onLongPress() {
    // @ts-expect-error ts-migrate(2551) FIXME: Property 'longPressTimeOut' does not exist on type... Remove this comment to see the full error message
    clearTimeout(this.state.longPressTimeOut);
    this.setState({longPressed: true, longPressTimeOut: null});
    (this.props as any).onLongPress();
  }
  onClick(e: any) {
    this.start(e);
  }
  render() {
    const {children} = this.props;
    const childrenWithProps = React.Children.map(children, (child) =>
      // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
      React.cloneElement(child, {
        onMouseDown: this.start,
        onTouchStart: this.start,
        onMouseUp: this.end,
        onTouchCancel: this.end,
        onTouchEnd: this.end,
        onClick: this.onClick,
      })
    );
    return childrenWithProps;
  }
}
export default OnLongPress;

/*
import {useCallback, useRef, useState} from "react";

const useLongPress = (
  onLongPress,
  onClick,
  {shouldPreventDefault = true, delay = 300} = {}
) => {
  const [longPressTriggered, setLongPressTriggered] = useState(false);
  const timeout = useRef();
  const target = useRef();

  const start = useCallback(
    (event) => {
      if (shouldPreventDefault && event.target) {
        event.target.addEventListener("touchend", preventDefault, {
          passive: false,
        });
        target.current = event.target;
      }
      timeout.current = setTimeout(() => {
        onLongPress(event);
        setLongPressTriggered(true);
      }, delay);
    },
    [onLongPress, delay, shouldPreventDefault]
  );

  const clear = useCallback(
    (event, shouldTriggerClick = true) => {
      timeout.current && clearTimeout(timeout.current);
      shouldTriggerClick && !longPressTriggered && onClick();
      setLongPressTriggered(false);
      if (shouldPreventDefault && target.current) {
        target.current.removeEventListener("touchend", preventDefault);
      }
    },
    [shouldPreventDefault, onClick, longPressTriggered]
  );

  return {
    onMouseDown: (e) => start(e),
    onTouchStart: (e) => start(e),
    onMouseUp: (e) => clear(e),
    onMouseLeave: (e) => clear(e, false),
    onTouchEnd: (e) => clear(e),
  };
};

const isTouchEvent = (event: any) => {
  return "touches" in event;
};

const preventDefault = (event: any) => {
  if (!isTouchEvent(event)) return;

  if (event.touches.length < 2 && event.preventDefault) {
    event.preventDefault();
  }
};
 * */
