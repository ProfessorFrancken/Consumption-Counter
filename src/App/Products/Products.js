import React from "react";
import PropTypes from "prop-types";
import Product from "./Product";

const Category = ({category, onClick, toggle, name, locked}) => (
  <nav className={"categoryRow"}>
    {category.map((product) => (
      <Product
        product={product}
        onClick={onClick}
        toggle={toggle}
        key={product.id}
        locked={product.locked || locked}
      />
    ))}
  </nav>
);

const Products = ({products, addProductToOrder, toggle}) => {
  const beer = products["Bier"] || [];
  const drinks = products["Fris"] || [];
  const food = products["Eten"] || [];

  return (
    <div className="productsGrid">
      {beer.length > 0 && (
        <Category
          onClick={addProductToOrder}
          toggle={toggle}
          category={beer}
          locked={false}
        />
      )}
      {drinks.length > 0 && (
        <Category onClick={addProductToOrder} toggle={toggle} category={drinks} />
      )}
      {food.length > 0 && (
        <Category onClick={addProductToOrder} toggle={toggle} category={food} />
      )}
    </div>
  );
};

const ProductPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
});

Products.propTypes = {
  products: PropTypes.shape({
    Bier: PropTypes.arrayOf(ProductPropType).isRequired,
    Fris: PropTypes.arrayOf(ProductPropType).isRequired,
    Eten: PropTypes.arrayOf(ProductPropType).isRequired,
  }).isRequired,
  addProductToOrder: PropTypes.func.isRequired,
};

export default Products;
