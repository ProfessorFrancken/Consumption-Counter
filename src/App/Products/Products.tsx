import React from "react";
import {
  Product as ProductType,
  useOrder,
  useOrderableProducts,
  useSelectedMember,
} from "./OrdersContext";
import Product from "./Product";

const Category = ({
  products,
  onClick,
  onLongPress,
  name,
}: {
  products: ProductType[];
  onClick: (product: ProductType) => void;
  onLongPress: (product: ProductType) => void;
  name: string;
}) => {
  return (
    <nav className={"categoryRow"} aria-label={`${name} category`}>
      {products.map((product) => {
        return (
          <Product
            key={product.id}
            product={product}
            onClick={() => onClick(product)}
            onLongPress={() => onLongPress(product)}
          />
        );
      })}
    </nav>
  );
};

const Products = () => {
  const {order, makeOrderMutation, addProductToOrder} = useOrder();
  const products = useOrderableProducts();

  const beer = products["Bier"] || [];
  const soda = products["Fris"] || [];
  const food = products["Eten"] || [];

  const hasPlacedATeporaryOrder = order.products.length > 0;

  const handleClick = (product: ProductType) => {
    if (hasPlacedATeporaryOrder) {
      addProductToOrder(product);
      return;
    }

    makeOrderMutation.mutate({products: [product]});
  };

  return (
    <div className="productsGrid">
      {beer.length > 0 && (
        <Category
          name="beer"
          onClick={handleClick}
          onLongPress={addProductToOrder}
          products={beer}
        />
      )}
      {soda.length > 0 && (
        <Category
          name="soda"
          onClick={handleClick}
          onLongPress={addProductToOrder}
          products={soda}
        />
      )}
      {food.length > 0 && (
        <Category
          name="food"
          onClick={handleClick}
          onLongPress={addProductToOrder}
          products={food}
        />
      )}
    </div>
  );
};

export default Products;
