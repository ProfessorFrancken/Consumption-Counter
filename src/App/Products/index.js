import { connect } from 'react-redux';
import { addProductToOrder } from '../../actions';
import { productsWithOrderCountSelector } from './../../selectors';
import { buyMore } from '../../actions';
import Products from './Products';

const mapStateToProps = state => ({
  products: productsWithOrderCountSelector(state, {
    hour: new Date().getHours()
  })
});

const mapDispatchToProps = dispatch => ({
  addProductToOrder: product => dispatch(addProductToOrder(product)),
  toggle: product => dispatch(buyMore(product))
});

export default connect(mapStateToProps, mapDispatchToProps)(Products);
