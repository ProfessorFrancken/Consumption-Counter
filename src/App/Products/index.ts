import {connect} from "react-redux";
import {addProductToOrder, buyMore} from "actions";
import {productsWithOrderCountSelector} from "./selectors";
import Products from "./Products";

const mapStateToProps = (state: any) => ({
  products: productsWithOrderCountSelector(state, {
    hour: new Date().getHours(),
  }),
});

const mapDispatchToProps = (dispatch: any) => ({
  addProductToOrder: (product: any) => dispatch(addProductToOrder(product)),
  toggle: (product: any) => dispatch(buyMore(product)),
});

// @ts-expect-error ts-migrate(2345) FIXME: Argument of type '({ products, addProductToOrder, ... Remove this comment to see the full error message
export default connect(mapStateToProps, mapDispatchToProps)(Products);
