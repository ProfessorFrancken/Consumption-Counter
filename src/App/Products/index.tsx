import Products from "./Products";
import {useOrder} from "./OrdersContext";

const ProductsScreen = () => {
  const {products, addProductToOrder, addProductToOrderOrMakeOrder} = useOrder();

  return (
    <Products
      products={products}
      addProductToOrder={addProductToOrder}
      addProductToOrderOrMakeOrder={addProductToOrderOrMakeOrder}
    />
  );
};

export default ProductsScreen;
