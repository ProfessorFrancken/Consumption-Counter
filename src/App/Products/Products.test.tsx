import React from "react";
import {render, fireEvent} from "test-utils";
import Products from "./Products";

it("renders without crashing", () => {
  render(
    <Products
      products={{Bier: [], Fris: [], Eten: []}}
      addProductToOrder={(click) => click}
    />
  );
});

it("adds products to an order when clicked", async () => {
  const addToOrder = jest.fn();

  const products = (
    <Products
      products={{
        Bier: [{id: 1, name: "Hertog-Jan", image: ""}],
        Fris: [],
        Eten: [],
      }}
      addProductToOrder={addToOrder}
    />
  );

  const screen = render(products);

  // Since this component uses a long press event, we need to simulate mouse down and up
  fireEvent.mouseDown(screen.getByLabelText("Buy Hertog-Jan"));
  fireEvent.mouseUp(screen.getByLabelText("Buy Hertog-Jan"));

  expect(addToOrder).toBeCalledWith({id: 1, name: "Hertog-Jan", image: ""});
});

it("renders products from beer, drinks and food categories", () => {
  const products = (
    <Products
      products={{
        // @ts-expect-error ts-migrate(2322) FIXME: Type '{ id: number; name: string; image: string; a... Remove this comment to see the full error message
        Bier: [{id: 1, name: "Hertog-Jan", image: "", age_restriction: 0}],
        Fris: [],
        Eten: [],
      }}
      addProductToOrder={jest.fn()}
    />
  );

  expect(products).toMatchSnapshot();
});
