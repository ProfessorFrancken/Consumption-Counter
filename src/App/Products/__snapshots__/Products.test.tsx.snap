// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders products from beer, drinks and food categories 1`] = `
<Products
  addProductToOrder={[MockFunction]}
  products={
    Object {
      "Bier": Array [
        Object {
          "age_restriction": 0,
          "id": 1,
          "image": "",
          "name": "Hertog-Jan",
        },
      ],
      "Eten": Array [],
      "Fris": Array [],
    }
  }
/>
`;
